src/connector/src/sink/mod.rs

```diff
@@ -202,6 +205,44 @@ pub trait SinkWriterV1: Send + 'static {
     async fn abort(&mut self) -> Result<()>;
 }
 
+/// A free-form sink that may output in multiple formats and encodings. Examples include kafka,
+/// kinesis, nats and redis.
+///
+/// The implementor specifies required key & value type (likely string or bytes), as well as how to
+/// write a single pair. The provided `write_chunk` method would handle the interaction with a
+/// `SinkFormatter`.
+///
+/// Currently kafka takes `&mut self` while kinesis takes `&self`. So we use `&mut self` in trait
+/// but implement it for `&Kinesis`. This allows us to hold `&mut &Kinesis` and `&Kinesis`
+/// simultaneously, preventing the schema clone issue propagating from kafka to kinesis.
+pub trait FormattedSink {
```

Comment:
Will we have a `FormattedSinkWriter` that wraps a sink formatter and a formatted sink, and implements `SinkWriter` with the `write_batch` as the current `write_chunk`? In that way we can implement this trait for `KinesisSinkWriter` directly rather than its reference, and we don't have to create a new formatter for each write_batch call.

Reply:
I am not very familiar with existing organization of sinks (`Sink` vs `SinkWriter` vs `SinkWriterV1`). I only focused on the format and encode part. This new trait is only there to simplify how a `SinkFormatter` is used. There may be opportunities to consolidate `write_chunk` with `write_batch`, or `write_one` with `put_record`/`write_inner`, but I do not have any idea right now.

> and we don't have to create a new formatter for each write_batch call

I also think the formatter should only be created once. I am just keeping the existing behavior - an async stream is created for each `write_batch` call. This is likely to be fixed later when I add the construction of more formatters, for example append-only-proto, append-only-avro.